tensor_parallel.py

```diff
@@ -0,0 +1,315 @@
+                tensor_list[rank] = param_m2     
+                dist.all_gather(tensor_list, param_m2, group=process_group)
+                # Note: torch.cat already creates a contiguous tensor.
+                first_dim = param_m2.shape[0]
+                param_m2 = torch.cat(tensor_list, dim = 0).contiguous()
+            if "o_proj" in name or "down_proj" in name:
+                # cancat the splited param_m2.grad
+                process_group = dist.group.WORLD
+                tensor_list = [torch.empty_like(param_m2) for _ in range(dist.get_world_size())]
+                rank = dist.get_rank()
+                tensor_list[rank] = param_m2     
+                dist.all_gather(tensor_list, param_m2, group=process_group)
+                # Note: torch.cat already creates a contiguous tensor.
+                first_dim = param_m2.shape[0]
+                param_m2 = torch.cat(tensor_list, dim = -1).contiguous()
+                
+        testcase.assertEqual(param_m2, param_m1)
+ 
+#before q, k, v_proj
+class InputTpLinear(torch.autograd.Function):
+    """Pass the input to the model parallel region."""
+
+    @staticmethod
+    def symbolic(graph, input_):
+        return input_
+
+    @staticmethod
+    def forward(ctx, input_):
+        return input_
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        process_group = dist.group.WORLD
+        if not grad_output.is_contiguous():
+            grad_output = grad_output.contiguous()
+        dist.all_reduce(grad_output, op=dist.ReduceOp.SUM, group=process_group)
+        return grad_output 
+
+def input_tp_linear(input):
+    return InputTpLinear.apply(input)
+
+# for the q k v_proj output
+class ColTpLinear(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, output):
+        #split qkv_weight have already done in get_tensor_sharded_model() 
+        
+        return output
+        
+    @staticmethod
+    def backward(ctx, grad):
+        #all_gather the grad, [384, 768] -> [768, 768]
+        print(grad)
+        print(grad.shape)
+        process_group = dist.group.WORLD
+        tensor_list = [torch.empty_like(grad) for _ in range(dist.get_world_size())]
+        rank = dist.get_rank()
+        tensor_list[rank] = grad     
+        dist.all_gather(tensor_list, grad, group=process_group)
+        # Note: torch.cat already creates a contiguous tensor.
+        first_dim = grad.shape[0]
+        grad = torch.cat(tensor_list, dim=first_dim).contiguous()
+        return grad
+
+def col_tp_linear(output):
+    return ColTpLinear.apply(output)  
+  
+# for the o_proj output
+class RowTpLinear(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, output):
+        # Use c10d collectselfops
+        process_group = dist.group.WORLD
+        if not output[0].is_contiguous():
+            output[0] = output[0].contiguous()
+        dist.all_reduce(output[0], op=dist.ReduceOp.SUM, group=process_group)
+        return output[0]
```

Comment:
Really take care of your naming, `Linear` means a matrix multiplication, but here this class is only a communication operation.